<launch>

    <arg name="drone_id" value="0"/>
    <arg name="odom_topic" value="viscar/odom" />
    
    <!-- map -->
    <!--arg name="map_size_x_" value="42.0"/>
    <arg name="map_size_y_" value="40.0"/>
    <arg name="map_size_z_" value="5.0"/ -->

    <!-- random map -->
    <!--node pkg ="map_generator" name ="random_forest" type ="random_forest" output = "screen">    
        <param name="map/x_size"     value="35" />
        <param name="map/y_size"     value="$(arg map_size_y_)" />
        <param name="map/z_size"     value="$(arg map_size_z_)" />
        <param name="map/resolution" value="0.1"/>        

        <param name="ObstacleShape/seed" value="1"/>
        
        <param name="map/obs_num"    value="100"/>
        <param name="ObstacleShape/lower_rad" value="1.0"/>
        <param name="ObstacleShape/upper_rad" value="1.3"/>
        <param name="ObstacleShape/lower_hei" value="0.5"/>
        <param name="ObstacleShape/upper_hei" value="3.0"/>        

        <param name="map/circle_num" value="40"/>        
        <param name="ObstacleShape/radius_l" value="0.7"/>        
        <param name="ObstacleShape/radius_h" value="0.5"/>        
        <param name="ObstacleShape/z_l" value="0.7"/>        
        <param name="ObstacleShape/z_h" value="0.8"/>        
        <param name="ObstacleShape/theta" value="0.5"/>        

        <param name="sensing/radius" value="5.0"/>        
        <param name="sensing/rate"   value="1.0"/>  
        
        <param name="min_distance" value="1.2"/>         
        <remap from="/map_generator/global_cloud" to="/global_map"/>

    </node-->

    <!-- main algorithm params -->
    <!--include file="$(find car_planner)/launch/advanced_param.xml">
        <arg name="drone_id" value="$(arg drone_id)"/>
        <arg name="map_size_x_" value="$(arg map_size_x_)"/>
        <arg name="map_size_y_" value="$(arg map_size_y_)"/>
        <arg name="map_size_z_" value="$(arg map_size_z_)"/>
        <arg name="odometry_topic" value="$(arg odom_topic)"/>
        <arg name="camera_pose_topic" value="pcl_render_node/camera_pose"/>
        <arg name="depth_topic" value="pcl_render_node/depth"/>
        <arg name="cloud_topic" value="pcl_render_node/cloud"/-->

        <!-- intrinsic params of the depth camera -->
        <!--arg name="cx" value="321.04638671875"/>
        <arg name="cy" value="243.44969177246094"/>
        <arg name="fx" value="387.229248046875"/>
        <arg name="fy" value="387.229248046875"/-->
        <!-- maximum velocity and acceleration the drone will reach -->
        <!--arg name="max_vel" value="2.0" />
        <arg name="max_acc" value="6.0" /-->
        <!--always set to 1.5 times grater than sensing horizen-->
        <!--arg name="planning_horizon" value="7.5" />
        <arg name="use_distinctive_trajs" value="true" /-->
        <!-- 1: use 2D Nav Goal to select goal  -->
        <!-- 2: use global waypoints below  -->
        <!--arg name="flight_type" value="1" /-->
        <!-- global waypoints -->
        <!-- It generates a piecewise min-snap traj passing all waypoints -->
        <!--arg name="point_num" value="1" />
        <arg name="point0_x" value="20" />
        <arg name="point0_y" value="-20" />
        <arg name="point0_z" value="0" />
    </include-->

    <!-- trajectory server -->
    <node pkg="ugv_planner" name="ugv_traj_server" type="traj_server" output="screen">
      <remap from="~controller_cmd" to="/ugv_planning/wheels_vel_cmd"/>
      <remap from="~trajectory_topic" to="/ugv_planner_node/trajectory"/>
      <remap from="~odom" to="$(arg odom_topic)"/>
    </node>

    <!-- laser detect -->
    <!--node pkg="grid_map" name="laser_detect" type="laser_detect_node" output="screen"-->
    <node pkg="detector" name="laser_pointer" type="laser_pointer_node" output="screen">
      <remap from="~raw_image" to="/camera/color/image_raw" />
      <remap from="~target" to="/move_base_simple/goal" />
      <remap from="~odometry" to="/viscar/odom" />
    </node>


    <!--main planner-->
    <node pkg="ugv_planner" name="ugv_planner_node" type="ugv_planner_node" output="screen">
      <remap from="~trajectory_topic" to="/ugv_planner_node/trajectory"/>
      <remap from="global_map" to="/global_map"/>
      <remap from="odom" to="/viscar/odom"/>
    </node>
    <!-- use simulator -->

    <include file="$(find ugv_planner)/launch/simulator.xml">
        <arg name="drone_id" value="$(arg drone_id)"/>
        <arg name="map_size_x_" value="40"/>
        <arg name="map_size_y_" value="60"/>
        <arg name="map_size_z_" value="10"/>
        <arg name="init_x_" value="1.0"/>
        <arg name="init_y_" value="-1.0"/>
        <arg name="init_z_" value="0.0"/>
        <arg name="odometry_topic" value="$(arg odom_topic)" />
    </include>

    <!--simulate global map-->
    <include file="$(find globalmap_generator)/launch/tmap_gene.launch"/>
    <!-- rviz -->
    <node name="rviz" pkg="rviz" type="rviz" args="-d $(find ugv_planner)/param/default.rviz" required="true" />
</launch>
